src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestHeaderValueResolver.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Attribute\MapRequestHeader;
+use Symfony\Component\HttpKernel\Attribute\MapRequestHeaders;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Serializer\Exception\PartialDenormalizationException;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\ConstraintViolationList;
+use Symfony\Component\Validator\Exception\ValidationFailedException;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+class RequestHeaderValueResolver implements ValueResolverInterface
+{
+    /**
+     * @see DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS
+     */
+    private const CONTEXT_DESERIALIZE = [
+        'collect_denormalization_errors' => true,
+    ];
+
+    public function __construct(
+        private readonly SerializerInterface&DenormalizerInterface $serializer,
+        private readonly ?ValidatorInterface $validator = null,
+    ) {
+    }
+
+    public function resolve(Request $request, ArgumentMetadata $argument): iterable
+    {
+        $attribute = $argument->getAttributesOfType(MapRequestHeader::class, ArgumentMetadata::IS_INSTANCEOF)[0]
+            ?? $argument->getAttributesOfType(MapRequestHeaders::class, ArgumentMetadata::IS_INSTANCEOF)[0]
+            ?? null;
+
+        if (!$attribute) {
+            return [];
+        }
+
+        $headers = [];
+        $requestHeaders = $request->headers->all();
+
+        array_walk($requestHeaders, function ($value, $key) use (&$headers) {
+            $headers[$key] = implode(',', $value);
```

Comment:
To get an array from
![image](https://github.com/user-attachments/assets/2914a0b4-4f69-48eb-a8de-5e643ebcf7b8)
To
![image](https://github.com/user-attachments/assets/e408c162-5af2-4878-90f0-164d2a9a4f2f)

If not modified all of the properties in your DTO object need to be as array type or denormalization will failed.

Reply:
I see :thinking: 

there was 2 parts in my comment, "this should be done only on the requested header" was about avoiding doing this on all headers when we only need one

and the second part is the part you answered.

it's true that for DTO objects, it would be annoying to have arrays everywhere. But for the odd case where you want an array it's a bit annoying to have to split the string just for that. especially since headers can have commas in their value, it's not just using `explode()`.

I don't have a very good solution for this but I do have a suggestion, we could add arguments to `MapRequestHeader(s)`, that could look like this:

* `#[MapRequestHeaders(toStringDefault: "concat")` the default: concat each header values with commas (`""` when absent?)
* `#[MapRequestHeaders(toStringDefault: "first")` take the first value of header values, `null` when absent
* `#[MapRequestHeaders(toString: ["Date" => first])` concat the values for all headers except for Date: take the first value and ignore the rest
* `#[MapRequestHeaders(toString: ["Date" => "first"])` concat the values for all headers except for Date: take the first value and ignore the rest if any
* `#[MapRequestHeaders(toString: ["Date" => false])` concat the values for all headers except for Date: keep it an array

* `#[MapRequestHeader(toString: "concat")` the default: concat each header values with commas (`""` when absent?)
* `#[MapRequestHeader(toString: "first")` take the first value of header values, `null` when absent

This should make the default behavior good enough in most cases but gives the option to opt-out of the concat shortcut when needed

wdyt?